File: programming/8.2/main.cpp
#include <iostream>
#include <string>

#define M 5
#define N 10

using namespace std;

int randint(int min, int max) {
  return rand() % (max + 1 - min) + min;
}

void fillData(int **data) {
  for (int i = 0; i < M; i++) {
    for (int j = 0; j < N; j++) {
      data[i][j] = randint(0, 100);
    }
  }
}

void showData(int **data) {
  for (int i = 0; i < M; i++) {
    for (int j = 0; j < N; j++) {
      cout << data[i][j] << " ";
    }
    cout << endl;
  }
}

int  main() {
  srand(time(NULL));
  int **numbers1 = new int*[M];
  int **numbers2 = new int*[M];
  int **numbers3 = new int*[M];
  int **numbers4 = new int*[M];

  for (int i = 0; i < M; i++) {
    numbers1[i] = new int[N];
    numbers2[i] = new int[N];
    numbers3[i] = new int[N];
    numbers4[i] = new int[N];
  }
  
  fillData(numbers1);
  fillData(numbers2);
  fillData(numbers3);
  fillData(numbers4);

  cout << "\nNumbers: 1\n";
  showData(numbers1);
  cout << "\nNumbers: 2\n";
  showData(numbers2);
  cout << "\nNumbers: 3\n";
  showData(numbers3);
  cout << "\nNumbers: 4\n";
  showData(numbers4);

  // has all
  int hasAll[101];

  for (int i = 0; i < 101; i++) {
    hasAll[i] = 0;
  }

  for (int i = 0; i < M; i++) {
    for (int j = 0; j < N; j++) {
      hasAll[numbers1[i][j]] += 1;
      hasAll[numbers2[i][j]] += 1;
      hasAll[numbers3[i][j]] += 1;
      hasAll[numbers4[i][j]] += 1;
    }
  }

  cout << endl << "Has all:" << endl;

  for (int i = 0; i < 101; i++) {
    if (hasAll[i] >= 4) {
      cout << i << " ";
    }
  }

  cout << endl << endl;

  // has no one
  bool hasNoOne[101];

  for (int i = 0; i < 101; i++) {
    hasNoOne[i] = false;
  }

  for (int i = 0; i < M; i++) {
    for (int j = 0; j < N; j++) {
      hasAll[numbers1[i][j]] = true;
      hasAll[numbers2[i][j]] = true;
      hasAll[numbers3[i][j]] = true;
      hasAll[numbers4[i][j]] = true;
    }
  }

  cout << "Has no one:" << endl;

  for (int i = 0; i < 101; i++) {
    if (!hasNoOne[i]) {
      cout << i << " ";
    }
  }

  cout << endl << endl;

  // has in second and third
  int hasFew[101];

  for (int i = 0; i < 101; i++) {
    hasNoOne[i] = 0;
  }

  for (int i = 0; i < M; i++) {
    for (int j = 0; j < N; j++) {
      hasFew[numbers2[i][j]] += 1;
      hasFew[numbers4[i][j]] += 1;
    }
  }

  cout << "Has few 2 and 4: " << endl;

  for (int i = 0; i < 101; i++) {
    if (hasFew[i] >= 2) {
      cout << i << " ";
    }
  }

  cout << endl;

  return 0;
}

File: programming/8.2/main.py
from random import randint

def generateData():
  return [[randint(0, 100) for j in range(10)] for i in range(5)]

def showData(data):
  for i in data:
    print(i) 

numbers1 = generateData()
numbers2 = generateData()
numbers3 = generateData()
numbers4 = generateData()

# show all arrays
print('Numbers 1:')
showData(numbers1)
print('Numbers 2:')
showData(numbers2)
print('Numbers 3:')
showData(numbers3)
print('Numbers 4:')
showData(numbers4)

print()

# has all
hasAll = [0 for i in range(102)]

for i in range(5):
  for j in range(10):
    hasAll[numbers1[i][j]] += 1
    hasAll[numbers2[i][j]] += 1
    hasAll[numbers3[i][j]] += 1
    hasAll[numbers4[i][j]] += 1

print("Has all:")

for i in range(102):
  if (hasAll[i] >= 4):
    print(hasAll[i], end=" ")

print()

# has no one
hasNoOne = [True for i in range(102)]

for i in range(5):
  for j in range(10):
    hasNoOne[numbers1[i][j]] = False
    hasNoOne[numbers2[i][j]] = False
    hasNoOne[numbers3[i][j]] = False
    hasNoOne[numbers4[i][j]] = False

print("Has no one:")

for i in range(102):
  if hasNoOne[i]:
    print(i, end=" ")

print()

# has 2 and 4
hasFew = [0 for i in range(102)]

for i in range(5):
  for j in range(10):
    hasFew[numbers2[i][j]] += 1
    hasFew[numbers4[i][j]] += 1

print("Has 2 and 4:")

for i in range(102):
  if (hasFew[i] >= 2):
    print(i, end=" ")